adding Doc on retrieving Image SBOMs (#1149)
chainguard-dev/edu: content/chainguard/chainguard-images/retrieve-image-sboms.md

[Cosign](/open-source/sigstore/cosign/an-introduction-to-cosign/) — a part of the Sigstore project — supports software artifact signing, verification, and storage in an [OCI (Open Container Initiative)](/open-source/oci/what-is-the-oci/) registry, as well as the retrieval of said artifacts. This tutorial outlines how you can use the `cosign` command to retrieve a Chainguard Image's SBOM. 

## Prerequisites

In order to follow this guide, you'll need the following installed on your local machine:

* **Cosign** — to retrieve SBOMs associated with Chainguard Images, check out [our guide on installing Cosign](/open-source/sigstore/cosign/how-to-install-cosign/) to configure it.
* **jq** — to process JSON, visit the [jq downloads](https://jqlang.github.io/jq/download/) page to set it up.

## Using Cosign to retrieve an Image's SBOM

Cosign includes a `download` command that allows you to retrieve a Chainguard Image's attestation over the command line. To do so, you would use this command with syntax like the following.

```shell

> cosign download attestation \ 

Reviewer: for some reason on macos, I'm getting this error unless I put the command all on one line: 

```
cosign download attestation \ 
cgr.dev/chainguard/python | jq -r .payload | base64 -d | jq .predicate
Error: could not parse reference:  
main.go:74: error during command execution: could not parse reference:  
-bash: cgr.dev/chainguard/python: No such file or directory
```

but then:

```
cosign download attestation cgr.dev/chainguard/python | jq -r .payload | base64 -d | jq .predicate
{
  "accounts": {
...
```

Author: Thanks for double checking, I'll put it all on one line to be safe

cgr.dev/chainguard/$IMAGE | jq -r .payload | base64 -d | jq .predicate
```

> 

Reviewer: Could you include an example image here? It would just make it easier for someone to copy / paste

> Notice that this example syntax includes `download attestation` rather than `download sbom`. You can generally think of an attestation as a an authenticated statement about a software artifact. There are different types of attestations [as defined by the SLSA 1.0 specification](https://slsa.dev/attestation-model), and they are typically referenced by their **predicate type**. One of the available predicate types is SPDX, an open standard for SBOM files. Because attestations must be signed, this is a way to verify the authenticity of the software producer, thereby ensuring the accuracy of the SBOM and the quality of the software.

Reviewer: ```suggestion
Notice that this example syntax includes `download attestation` rather than `download sbom`. You can generally think of an attestation as an authenticated statement about a software artifact. There are different types of attestations [as defined by the SLSA 1.0 specification](https://slsa.dev/attestation-model), and they are typically referenced by their **predicate type**. One of the available predicate types is SPDX, an open standard for SBOM files. Because attestations must be signed, this is a way to verify the authenticity of the software producer, thereby ensuring the accuracy of the SBOM and the quality of the software.
```

As an example, to retrieve the `argocd` image's attestation you would run a command like this.

```shell
cosign download attestation \
  --platform=linux/amd64 \
  --predicate-type=https://spdx.dev/Document \
  cgr.dev/chainguard/argocd | jq -r .payload | base64 -d | jq .predicate
```

This example includes two extra arguments not included in the example syntax outlined previously. First, it includes the `--platform` flag which allows you to download the attestation for a specific platform image. This example specifies the `linux/amd64` platform, but you could also use `linux/arm64`. Be aware, though, that in order to use the `--platform` option you'll need to have Cosign version 2.2.1 or newer installed.

The other extra argument is the `--predicate-type` flag, required to specify which type of predicate you want to download from the registry. In order to download Chainguard Images SBOM attestations, you should use the `https://spdx.dev/Document` predicate type.

## Learn more

> We provide provenance information for every Chainguard Image in their respective [Reference docs](/chainguard/chainguard-images/reference/). After reaching the **Overview** for the image of your choice, navigate to the **Provenance** tab for information on how to retrieve the image's attestations, as well as how to verify the image's attestations and signatures. 

Reviewer: Can you include just an example of this...

> For example, if you're looking for the provenance information of the Python image, you can navigate to the [Python overview page](/chainguard/chainguard-images/reference/python/) and then click through to **Provenance** which will bring you to the following URL: [https://edu.chainguard.dev/chainguard/chainguard-images/reference/python/provenance_info/](https://edu.chainguard.dev/chainguard/chainguard-images/reference/python/provenance_info/).

I thought having the URL may help people replace the image they want to look for?

---
title: "How to Retrieve SBOMs for Chainguard Images"
linktitle: "Retrieve an Image's SBOM"
type: "article"
description: "A brief tutorial on how to use Cosign to retrieve Chainguard Image SBOMs."
date: 2023-11-17T11:07:52+02:00
lastmod: 2023-11-17T11:07:52+02:00
draft: false
tags: ["Conceptual", "Chainguard Images", "SBOM"]
images: []
menu:
  docs:
    parent: "chainguard-images"
weight: 710
toc: true
---

Chainguard Images contain only the minimum number of packages needed to use the software they contain. The purpose of this is to reduce the image's attack surface and minimize the risk that CVEs will impact software that depends on these container images. 

> Even though they contain the minimum number of packages, there may come a time when you want to know exactly what's running inside of a certain Chainguard Image. For this reason, we include a signed SBOM (also known as an *attestation*) with each image.

Reviewer 2: ```suggestion
Even though they contain the minimum number of packages, there may come a time when you want to know exactly what's running inside of a certain Chainguard Image. For this reason, we include a signed SBOM with each image in the form of a [software attestation](https://slsa.dev/attestation-model).
```

Reviewer 2: I think some people could read this and understand that an SBOM is an attestation, when in fact we offer our sboms as a slsa attestation but there are other types of attestations.

[Cosign](/open-source/sigstore/cosign/an-introduction-to-cosign/) — a part of the Sigstore project — supports software artifact signing, verification, and storage in an [OCI (Open Container Initiative)](/open-source/oci/what-is-the-oci/) registry, as well as the retrieval of said artifacts. This tutorial outlines how you can use the `cosign` command to retrieve a Chainguard Image's SBOM using Cosign. 

## Prerequisites

In order to follow this guide, you'll need the following installed on your local machine:

* **Cosign** — to retrieve SBOMs associated with Chainguard Images, check out [our guide on installing Cosign](/open-source/sigstore/cosign/how-to-install-cosign/) to configure it.
* **jq** — to process JSON, visit the [jq downloads](https://jqlang.github.io/jq/download/) page to set it up.

## Using Cosign to retrieve an Image's SBOM

> Cosign includes a `download` command that allows you to retrieve a Chainguard Image's SBOM over the command line. To do so, you would use this command with syntax like the following.

Reviewer 2: ```suggestion
Cosign includes a `download` command that allows you to retrieve a Chainguard Image's attestation over the command line. To do so, you would use this command with syntax like the following.
```

```shell
cosign download attestation \ 
cgr.dev/chainguard/$IMAGE | jq -r .payload | base64 -d | jq .predicate
```

> Notice that this example syntax includes `download attestation` rather than `download sbom`. While there are a few [differences between SBOMs and attestations](/open-source/sbom/sboms-and-attestations/), you can generally think of an attestation as a more accurate representation of the contents of the container image than an SBOM. This is because attestations must be signed by the software producer, thereby ensuring the accuracy of the SBOM and the quality of the software. 

Reviewer 2: ```suggestion
Notice that this example syntax includes `download attestation` rather than `download sbom`. You can generally think of an attestation as a an authenticated statement about a software artifact. There are different types of attestations [as defined by the SLSA 1.0 specification](https://slsa.dev/attestation-model), and they are typically referenced by their **predicate type**. One of the available predicate types is SPDX, an open standard for SBOM files. Because attestations must be signed, this is a form of verifying the  authenticity of the software producer, thereby ensuring the accuracy of the SBOM and the quality of the software. 
```

Reviewer 2: the attestation is in fact like an envelope that wraps the SBOM with additional metadata and is signed, so it can be verified. I hope this helps clarifying it a bit more, but feel free to move things around for better fluidity :)

Author: This is really helpful framing, thanks Erika!

This attestation data is encoded in base64, making it unreadable without further processing. 
This is why the output from the first part of the command is piped into `jq` in order to filter out the payload section of the output containing the SBOM.

This filtered output is then passed into the `base64` command to be decoded before that output is piped into another `jq` command. This final `jq` command extracts the attestation predicate from the `base64` output and returns it to your terminal.

As an example, to retrieve the `argocd` image's attestation you would run a command like the following.

```shell
cosign download attestation \
  --platform=linux/amd64 \
  --predicate-type=https://spdx.dev/Document \
  cgr.dev/chainguard/argocd | jq -r .payload | base64 -d | jq .predicate
```

> This example includes two extra arguments not included in the example syntax outlined previously. First, it includes the `--platform` flag which allows you to download the attestation for a specific platform image. This example specifies the `linux/amd64` platform, but you could enter alternatives such as `linux/arm64` or `darwin/amd64`. Be aware, though, that in order to use the `--platform` option you'll need to have Cosign version 2.2.1 or newer installed.

Reviewer 2: ```suggestion
This example includes two extra arguments not included in the example syntax outlined previously. First, it includes the `--platform` flag which allows you to download the attestation for a specific platform image. This example specifies the `linux/amd64` platform, but you could also use `linux/arm64`. Be aware, though, that in order to use the `--platform` option you'll need to have Cosign version 2.2.1 or newer installed.
```

Reviewer 2: AFAIK we only provide attestations for amd64 and arm64. "darwin/amd64" returns a not found error:

```
❯ cosign download attestation \
  --platform=darwin/amd64 \
  --predicate-type=https://spdx.dev/Document \
  cgr.dev/chainguard/argocd | jq -r .payload | base64 -d | jq .predicate
Error: unable to find an attestation for darwin/amd64
main.go:74: error during command execution: unable to find an attestation for darwin/amd64
```

Author: Ah that's my bad, thanks for catching that Erika!

> The other extra argument is the `--predicate-type` flag. This allows you to specify which format you'd like the SBOM you're downloading to follow. This example retrieves the SBOM in the SPDX format, but to download a CycloneDX SBOM you could instead have the `--predicate-type` option point to `https://cyclonedx.org/Document`.

Reviewer 2: ```suggestion
The other extra argument is the `--predicate-type` flag, required to specify which type of predicate you want to download from the registry. In order to download Chainguard Images SBOM attestations, you should use the `https://spdx.dev/Document` predicate type.
```

## Learn more

> We provide provenance information for every Chainguard Image in their respective [Reference docs](/chainguard/chainguard-images/reference/). After reaching the **Overview** for the image of your choice, navigate to the **Provenance** tab for information on how to retrieve the image's attestation, as well as how to verify the image's attestation and signatures. 

Reviewer 2: ```suggestion
We provide provenance information for every Chainguard Image in their respective [Reference docs](/chainguard/chainguard-images/reference/). After reaching the **Overview** for the image of your choice, navigate to the **Provenance** tab for information on how to retrieve the image's attestations, as well as how to verify the image's attestations and signatures. 
```
